WIP: Add JUnitUtil.setInstanceManager()

danielb987 opened the pull request:
This is a demonstration for #5869

danielb987 commented:
In the test LayoutEditorExtendedTest.testSomething(), the first Assert will succeed since the turnout "3" in the first "JMRI" setup is thrown.

But the second Assert will fail, since the turnout "3" in the second "JMRI" setup is closed.

pabender commented:
I really do believe what I said in one of my replies to #5869, which is that you're trying to test consistency here, which is not the same as testing correctness.

bobjacobsen commented:
Would it be possible to move the `getAllInstances` method to the java/test tree either in a new InstanceManagerUtil class, or as part of JUnitUtil?  As written, it seems to need to know too much about other specific classes, and I'd prefer to keep InstanceManager at the root of the reference tree.  (That's why the various type-specific calls at the bottom have been deprecated and are (slowly) being moved out.

Thanks for considering it.

bobjacobsen commented on java/src/jmri/InstanceManager.java:
```diff
@@ -771,6 +771,45 @@ public InstanceManager() {
+        for (Class<?> type : managerLists.keySet()) {
+            synchronized (type) {
+                for (Object manager : managerLists.get(type)) {
+                    if (manager instanceof jmri.Manager) {
```
I don't quite understand the `if` structure here.  If you're looking for jmri.Managers (which manage NamedBeans), why are the "Ignore this" clauses needed?

danielb987 replied:
I'm a bit in the dark here, so I added the `if` statement to ensure that I don't miss something. The important part is the `throw` statement at the bottom. If a new manager type is added later to the instance manager but this method doesn't know about it, the method will throw an exception.

But if it is absolute sure that there will never be a manager that isn't an instance of `jmri.Manager` that holds any bean that we are interested in, then that `if` statement can be removed. It is probably so, and if you agree with that I will remove the `if` statement. But I want to be sure before I remove it.

bobjacobsen commented on java/src/jmri/managers/DefaultSignalMastLogicManager.java:
```diff
@@ -330,7 +330,11 @@ public int getObjectCount() {
     /** {@inheritDoc} */
     @Override
     public SortedSet<SignalMastLogic> getNamedBeanSet(){
```
Could you add a comment mentioning that this is not a live set (i.e. it won't see later changes)?  Having this method here is progress, but eventually it needs to meet the contract of the parent type that this  set is live; that's slowly being built into other parts of the system.

danielb987 replied:
Yes, I can add the comment. But it would be good to fix the contract instead.

I got a NPE so I was forced to do something, but I don't fully understand what needs to be done. This was a quick fix to get a working demo.

How do I change this method to meet the contract?

danielb987 commented:
@bobjacobsen : Thanks for the review.

There are many things with JMRI that I don't fully understand and sometimes I feel like being in a dark room, trying to get forward. But it's very fun to learn and I appreciate the comments a lot.

I have moved InstanceManager.getAllInstances() to JUnitUtil, changed the List to a Set and removed the `@SuppressWarnings`.

bobjacobsen commented:
> The important part is the throw statement at the bottom. If a new manager type is added later to the instance manager but this method doesn't know about it, the method will throw an exception.
> 
> But if it is absolute sure that there will never be a manager that isn't an instance of jmri.Manager that holds any bean that we are interested in, then that if statement can be removed. 

Items of type <T extends jmri.NamedBean>, like Turnouts, Memories, etc, are stored and accessed via a jmri.Manager<T extends jmri.NamedBean> subclass. You should count on that.

But not everything with “Manager” in it’s name inherits from jmri.Manager.  You’ve got several listed, but  people can store just about anything they want in the Instance Manager.  That’s sort-of the point.  There’s no real constraint on that.  Since there’s not even a requirement of any specific `equals` implementation, you have to treat things that aren’t jmri.Managers as opaque types:  Just check that you still have the same number of objects of the same type.  The code could say something like:

“Both InstanceManagers have a list with N objects where .getClass().getName() says a.b.c so we call that a match and move on”

For example, “Both InstanceManagers have a list of two Throttle Manager objects where one claims to be an NceThrottleManager and the other is an InternalThrottleManager. Those two lists match, so I’ll move on to the next list"

The specific case of jmri.Manager is worth the special handling of asking for the named beans and checking the system name, user name, and state match.  

Bob

--
Bob Jacobsen
jacobsen@mac.com +1-510-708-5988 AIM, Skype JacobsenRG

bobjacobsen commented:
> On Sep 27, 2018, at 10:00 AM, Daniel <notifications@github.com> wrote:
> 
> Yes, I can add the comment. But it would be good to fix the contract instead.
> 
> I got a NPE so I was forced to do something, but I don't fully understand what needs to be done. This was a quick fix to get a working demo.
> 
> How do I change this method to meet the contract?

Easiest by far would be to have DefaultSignalMastLogicManager inherit from AbstractManager.  It’ll do what’s needed and inheriting will reduce replicated code.  

Alternately, change the internal storage from 

```
List<SignalMastLogic> signalMastLogic = new ArrayList<SignalMastLogic>()
```
to a Set, and then return a java.util.Collections#unmodifiableSet version of that set.  See the code in jmri.Manager around line 195.

danielb987 commented:
> But not everything with “Manager” in it’s name inherits from jmri.Manager. You’ve got several listed, but people can store just about anything they want in the Instance Manager. That’s sort-of the point. There’s no real constraint on that. Since there’s not even a requirement of any specific `equals` implementation, you have to treat things that aren’t jmri.Managers as opaque types: Just check that you still have the same number of objects of the same type. The code could say something like:
> 
> “Both InstanceManagers have a list with N objects where .getClass().getName() says a.b.c so we call that a match and move on”

In order to test that, the managers we want to test need to implement a interface that has a method `getNumberOfObjects()` . If that is desired, I will add that interface.

For example `ThrottleManager` doesn't have any method to get a list of the objects or the number of throttles it knows about.

danielb987 commented:
In my new commit, I have changed the method `getAllNamedBeansFromManagers` to only look at `jmri.Manager` and don't care about the rest of the managers. And I changed the name and the documentation of the method to reflect that. I'm not too happy about the new name of the method, `getAllNamedBeansFromManagerManagers()`, but I think the named should reflect that it's only getting named beans from `the jmri.Manager` managers.

bobjacobsen commented:
> In order to test that, the managers we want to test need to implement a interface that has a method getNumberOfObjects() . If that is desired, I will add that interface.
> 
> For example ThrottleManager doesn't have any method to get a list of the objects or the number of throttles it knows about.

Note that there can be more than one ThrottleManager; that’s what I was pointing out.

You can certainly make the contents of a throttle more visible, but I don’t know that all possible ThrottleManager implementations will be able to respond properly.  DCC systems are nasty that way.  I recommend you use special test ThrottleManager objects, i.e. from a ThrottleManagerScaffold class, if you need to do something special for testing.

pabender commented:
Sent from my iPad

> On Sep 27, 2018, at 1:20 PM, Daniel <notifications@github.com> wrote:
> 
> In order to test that, the managers we want to test need to implement a interface that has a method getNumberOfObjects() . If that is desired, I will add that interface.
> 
> For example ThrottleManager doesn't have any method to get a list of the objects or the number of throttles it knows about.
> 

This is true of the throttle manager interface.  AbstractThrottleManager is keeping a hash table with the throttles assigned to addresses.

I think most of the throttle managers extend AbstractThrottleManager.  The class diagram in the throttle manager interface javadoc certainly implies that.

http://jmri.sourceforge.net/JavaDoc/doc/jmri/ThrottleManager.html

( those class diagrams are automatically generated from the source.  If they aren’t right, we need to find out why )

pabender commented:
Sent from my iPad

> On Sep 27, 2018, at 1:36 PM, Bob Jacobsen <notifications@github.com> wrote:
> 
> You can certainly make the contents of a throttle more visible, but I don’t know that all possible ThrottleManager implementations will be able to respond properly. DCC systems are nasty that way. I recommend you use special test ThrottleManager objects, i.e. from a ThrottleManagerScaffold class, if you need to do something special for testing.

I think I have an example of doing just that because of the work I have been doing testing the throttle user interface.   I need to pull that out into a more visible location for another set of tests I am starting to work on.  I’ll get working on that in a few minutes.  PR to follow.

danielb987 commented:
> This is true of the throttle manager interface. AbstractThrottleManager is keeping a hash table with the throttles assigned to addresses.

Thanks for the note.

danielb987 commented:
@bobjacobsen : Huston, we have a problem.

In my last commit, I add the check that the two instance managers has the same number of managers of each type.

Please take a look at `JUnitUtilTest` and the methods `thisTestWorks()` and `thisTestFails()` . The difference between them is that the second test also calls `jmri.util.JUnitUtil.initLogixManager()` which the first test doesn't.

And as the method names say, the first test works and the second test fails. The reason is that `initLogixManager()` creates a `jmri.jmrit.logix.WarrantPreferences` for `layoutEditorInstanceManager` but not for `manualInstanceManager` .

It's fully possible that I'm doing some big misstake here, or making some bad assumption, but if I call `resetInstanceManager()` and then a number of `init????()` , shouldn't I get the exact same result each time?

danielb987 commented:
I added one more test, `JUnitUtilTest.testNumManagers()` .

The following code is executed five times in the same test:
```
jmri.util.JUnitUtil.resetInstanceManager();
jmri.util.JUnitUtil.initInternalTurnoutManager();
jmri.util.JUnitUtil.initInternalLightManager();
jmri.util.JUnitUtil.initInternalSensorManager();
jmri.util.JUnitUtil.initDebugThrottleManager();
jmri.util.JUnitUtil.initLogixManager();
jmri.util.JUnitUtil.initIdTagManager();
```
The first time, 21 managers is registered.
The second time, `jmri.jmrit.logix.WarrantPreferences` is missing.
The third time, both `jmri.jmrit.logix.WarrantPreferences` and `jmri.implementation.SignalSpeedMap` is missing.
The fourth and fifth times the number of registered managers stay the same.

Is this a bug in JMRI:s management of the managers or am I doing something wrong or making the wrong assumptions?

danielb987 commented:
I was able to solve the problem with `jmri.jmrit.logix.WarrantPreferences` by adding `resetProfileManager()` and `initConfigureManager()` .

But I have still the problem with `jmri.implementation.SignalSpeedMap` . The strange thing is that the first and the second time, `SignalSpeedMap` gets registered. But the third, fourth and fifth time it does not get registred in the instance manager.

danielb987 commented:
Travis still fail on my test, which expects 42 managers but only finds 41 managers at row 71 in `JUnitUtilTest.testNumManagers()` . The test works fine when I run it on my computer but not when Travis runs all the tests.

Is there some initialization that I have forgotten to do? Calling some init??() metod?

danielb987 commented:
I have added a test `JUnitUtilTest.testNumManagers2()` which always creates a SignalSpeedMap manager after the instance managers has been initialized. I want to see if that solves the problem.

pabender commented:
> I think I have an example of doing just that because of the work I have been doing testing the throttle user interface. I need to pull that out into a more visible location for another set of tests I am starting to work on. I’ll get working on that in a few minutes. PR to follow.

So it turns out I have been using the DebugThrottleManager and installing it via:

https://github.com/JMRI/JMRI/blob/a7094cd90925c2c9ca1ccbc6cc090cd3a952e8d9/java/test/jmri/util/JUnitUtil.java#L669-L672

I don't know that we need a separate throttle manager scaffold since we have that.

pabender commented:
> @pabender :
> If a method `AbstractThrottleManager.getNumThrottles()` is added, it will probably solve the issue with detecting new throttles, since `DebugThrottleManager` inherit `AbstractThrottleManager` .
> 
> I think that `AbstractThrottleManager.getNumThrottles()` should return `addressThrottles.values().size()` 

That sounds reasonable to me.

> since at least some systems allow several throttles use the same address. LocoNet supports two throttles controlling the same address. But maybe JMRI doesn't?

To the best of my knowledge, JMRI throttle managers create one throttle object per address.  The throttle object is the interface between JMRi and the system specific locomotive commands.  That throttle object can have multiple interfaces controlling it (scripts, the throttle UI, WiThrottle, etc) but there is only ever one throttle.

Technically, LocoNet doesn't allow two throttles to control a slot.  There is only ever one "in control" but more than one can fight over the slot.

bobjacobsen commented:
If you're only adding this `getNumThrottles()` method for testing, please don't.  It's going to be an ongoing pain in the neck to maintain it across all our systems, and we haven't had an operational need for it.  

Instead, please use a debug ThrottleManager, or cast down to the `AbstractThrottleManager` implementation.

bobjacobsen commented:
It looks to me like the WarrantPreferences code strongly interacts with the preferences stored in the profile.  

I don’t think that "that resetInstanceManager() not fully reset the managers and that some things is left which shouldn't be there.”  That’s possible, but it’s not likely.  Much more likely is that (a) things have state _outside_ the InstanceManager (i.e. in statics and/or the preferences system), or (b) you’re seeing the effects of threading on the automatic creation of some items. 

> On Sep 27, 2018, at 1:38 PM, Daniel <notifications@github.com> wrote:
> 
> The first time, 21 managers is registered.
> The second time, jmri.jmrit.logix.WarrantPreferences is missing.
> The third time, both jmri.jmrit.logix.WarrantPreferences and jmri.implementation.SignalSpeedMap is missing.
> The fourth and fifth times the number of registered managers stay the same.

danielb987 commented:
Thanks.

I'm starting to understand the InstanceManager class. I thought that `JUnitUtil.resetInstanceManager()` cleared out all the managers and that no manager was left. But I now realize that it's not always true.

`JUnitUtil.resetInstanceManager()` calls `InstanceManager.getDefault().clearAll()` which does:
```
LazyInstanceManager.instanceManager = new InstanceManager();
new HashSet<>(managerLists.keySet()).forEach((type) -> {
    clear(type);
});
```

The method clear() does:
```
List<T> toClear = new ArrayList<>(getInstances(type));
toClear.forEach((o) -> {
    remove(o, type);
});
```

The method remove() does:
```
List<T> l = getList(type);
int index = l.indexOf(item);
....
if (index != -1) { // -1 means items was not in list, and therefor, not registered
    // fire property change last
    pcs.fireIndexedPropertyChange(getListPropertyName(type), index, item, null);
}

```

And the method `pcs.fireIndexedPropertyChange()` in some cases put the manager back in the new InstanceManager.

Therefore there may be managers in the new InstanceManager after the call to `resetInstanceManager()` .

But if I call `resetInstanceManager()` twice, the managers will be cleared. So by calling `resetInstanceManager()` twice, I will solve the problem.

danielb987 commented:
What is the recommendation about existing calls to depricated methods? Fix them in a PR when I see them, or leave them until later? For example, `JUnitUtil.initInternalSignalHeadManager()` has a call to `InstanceManager.setSignalHeadManager()` . Same with `initDebugCommandStation()` and `initDebugProgrammerManager()` which I think should call `InstanceManager.store()` instead of depricated methods.

bobjacobsen commented:
> And the method pcs.fireIndexedPropertyChange() in some cases put the manager back in the new InstanceManager.
> 
> Therefore there may be managers in the new InstanceManager after the call to resetInstanceManager() .
> 

Interesting! It’s good you tracked that down. 

That might actually be a misuse that we should test for and fix. What’s an example of a manager that comes back?

Is this because of the auto-create, or is something manually adding a manager back?

Thanks again for taking the time to find this.

bobjacobsen commented:
> What is the recommendation about existing calls to depricated methods?

Particularly when the replacement is clear, I try to fix them when I’m working in the file. But I don’t always get every one...

danielb987 commented:
> That might actually be a misuse that we should test for and fix. What’s an example of a manager that comes back?

I have added `jmri.util.JUnitUtilAdditionalTest` . That test does a complete reset, then init one manager, which may or may not init other managers. Then I reset one time and check if the instance manager has got a complete reset.

If I have got it right, it seems that the following managers don't reset completely:

Light, SignalMastLogic, Logix, Turnout, SignalMast, SignalSpeedMap, Power, SignalHead,  Section, Configure, Route, Programmer, IdTag, Sensor, RailCom, CommandStation, Throttle, LayoutBlock, Shutdown and TurnoutOperation.

But Reporter, UserMessagesPreferences, Memory, OBlock, Warrant and Conditional seems to work.

> Is this because of the auto-create, or is something manually adding a manager back?

I don't have a clue.

danielb987 commented:
I also note that the tests doesn't give the same result on my local computer as it does on Travis. The manager SIGNAL_MAST_LOGIC creates a different number of managers on my local computer compared to Travis.

danielb987 commented:
> Working from phone for a bit longer, can't investigate in detail, but it appears that the InstanceManager default constructor (near line 750) creates _all_ the objects that are specified as services right at startup, instead of waiting for them to be requested.

The new `InstanceManager` is created first, and then managers are added to the new instance manager. And in order to be able to do that, the managers need to get the new manager from `InstanceManager.getDefault()` . I have therefore added a commit that prints a stack trace if `getDefault()` is called from the new instance mananger in the method `InstanceManager.clearAll()` .

It turns out that the method `InstanceManager.remove` , which is a non static method, calls the static method `InstanceManager.getList(type)` , which in turn calls `InstanceManager.getDefault()` to get the instance manager.

This means that the method `InstanceManager.getList(type)` returns the list of managers from the new instance manager instead of the old instance manager.

It seems to me that we need a non static method that do what `InstanceManager.getList(type)` does, but do it on the correct instance manager.

danielb987 commented:
I added the method `InstanceManager.getList(instanceManager, type)` and that solved the problem for all managers except `IdTagManager` . I don't know what makes this manager special.

I have added `JUnitUtilAdditionalTest.testIdTagManager()` that tests only `IdTagManager` , and this manager leave behind `ConfigureManager`, `ShutDownManager` and `UserPreferencesManager` .

danielb987 commented:
In the method `InstanceManager.clearAll()` , does the statement 
`LazyInstanceManager.instanceManager = new InstanceManager();`
needs to be at the beginning? If it's moved later down in the method, the issues with `AbstractManager.dispose()` and `DefaultIdTagManager.dispose()` will be solved.

Is there anything in `clearAll()` that needs the new instance manager? If not, I think that the new instance manager should be created last in `clearAll()` .

danielb987 commented:
> DefaultIdTagManager.dispose() does:
> ....
> 
> And `AbstractManager.dispose()` does:
> ....

The difference between IdTagManager and the other managers is that `IdTagManager` inherits `Disposable` and therefore gets disposed. But it seems to be the only manager that inherits `Disposable` and therefore the other managers don't get disposed.

Many of the `init?????Manager()` methods call `registerConfig()` on the registered `ConfigureManager` but since the managers don't get disposed, `AbstractManager.dispose()` is not called and therefore they are not deregistered from the configure manager.

Shouldn't the other managers also get disposed so that `AbstractManager.dispose()` is called and so they get deregistered from the configure manager?

What is the reason to not let `AbstractManager` inherit `Disposable` to ensure that all it's sub classes get disposed?

danielb987 commented:
The following managers inherits `AbstractManager` but doesn't inherit `Disposable` :

```
jmri.BlockManager
jmri.jmrit.display.layoutEditor.LayoutBlockManager
jmri.jmrit.logix.OBlockManager
jmri.jmrit.logix.WarrantManager
jmri.managers.AbstractSignalHeadManager
jmri.managers.DefaultConditionalManager
jmri.managers.DefaultMemoryManager
jmri.managers.DefaultRailComManager
jmri.managers.DefaultRouteManager
jmri.managers.DefaultSignalMastManager
```

danielb987 commented:
@bobjacobsen 

I tried to merge upstream master into this PR but it failed miserably. Maybe one could undo the merge into branch `danielb987:layouteditor3` . However, there is a bigger problem.

The intention with InstanceManager.clearAll() is to get a new InstanceManager that is clean. However, it turns out that managers in the old InstanceManager pollutes the new InstanceManager that is created in clearAll() at line 808.

To try to solve this, I created a new branch from master there I tried to move the line
`LazyInstanceManager.instanceManager = new InstanceManager();`
to the end of the method `InstanceManager.clearAll()`. That way, the managers in the old InstanceManager would not be able to pollute the new InstanceManager. But the result was a lot of warnings
`list of interface jmri.ConfigureManager was reinitialized during clearAll`.

Instead of moving the creation of a new InstanceManager to the end of the method, one could fix the problem with the method `InstanceManager.remove()` calls the static method `getList()`. But there is at least one class that calls the new InstanceManager in its dispose() method and that is IdTagManager with its parent DefaultIdTagManager.

The problem with DefaultIdTagManager.dispose() is that it calls InstanceManager.getOptionalDefault and that method does not do what it is intended to do in this particular case. The intent is to check if a ShutDownManager is present and if so deregister a task.

But getOptionalDefault calls getNullableDefault() which calls getInstance() which in some cases creates a new manager instead of returning null.

I think this is too complicated for me so I close this PR and PR #5907.

bobjacobsen commented:
> 
> Closed #5881.


Don’t delete the branch, please, I’d like to take a look at it.

Unfortunately, I’m _way_ behind on getting the Internal fix done, so it will take a while before I can get to this.

Bob

danielb987 commented:
I will keep the branch for you.
